roachtest: acceptance/gossip/peerings failed [fix incoming]

SHA: https://github.com/cockroachdb/cockroach/commits/32346197937265bbc71dd3629b1137df9bf16fb4

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=acceptance/gossip/peerings PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=948285&tab=buildLog

```
The test failed on master:
	test.go:500,gossip.go:260,acceptance.go:61: failed to get gossip status from node 2: Get http://35.225.217.78:26258/_status/gossip/local: dial tcp 35.225.217.78:26258: connect: connection refused

```

Node 2 crashed:

```
E181007 05:35:33.085080 135 util/log/crash_reporting.go:477  [n3,s4] Reported as error 1ad9e478756449f399d5fe45c4183b1b
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 135 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420532750, 0x30a22c0, 0xc42025e450)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:184 +0x11f
panic(0x27ca500, 0x42c9db0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/cockroachdb/cockroach/pkg/storage.DecodeRaftCommand(0x0, 0x0, 0x0, 0x2, 0x1, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_raftstorage.go:1059 +0x19e
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).maybeTrackForwardedProposalLocked(0xc420bfc300, 0xc42591a880, 0xc42547a188)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4038 +0x217
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).stepRaftGroup.func1(0xc42591a880, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3981 +0x197
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked.func1(0xc425e0db88, 0xd, 0xc42591a880, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:575 +0x30
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroupLocked(0xc420bfc300, 0x2c98100, 0xc425e0db88, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:576 +0xc2
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).withRaftGroup(0xc420bfc300, 0x0, 0xc425e0db88, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:601 +0x8c
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).stepRaftGroup(0xc420bfc300, 0xc42547a188, 0x0, 0xc4258277c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3962 +0x55
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequestWithReplica(0xc4204dfb80, 0x30a22c0, 0xc425928cf0, 0xc420bfc300, 0xc42547a188, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3555 +0xeb
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue.func1(0x30a22c0, 0xc425928cf0, 0xc420bfc300, 0x30a22c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3861 +0x97
github.com/cockroachdb/cockroach/pkg/storage.(*Store).withReplicaForRequest(0xc4204dfb80, 0x30a22c0, 0xc425928cf0, 0xc42547a188, 0xc425e0ded0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3518 +0x135
github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRequestQueue(0xc4204dfb80, 0x30a22c0, 0xc42025e450, 0xd)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3855 +0x229
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker(0xc42047f400, 0x30a22c0, 0xc42025e450)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:225 +0x21b
github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2(0x30a22c0, 0xc42025e450)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:165 +0x3e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc4202c3750, 0xc420532750, 0xc4202c3740)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:199 +0xe9
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
```

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -4013,6 +4013,13 @@ func (r *Replica) shouldDropForwardedProposalLocked(req *RaftMessageRequest) boo
 	for _, e := range req.Message.Entries {
 		switch e.Type {
 		case raftpb.EntryNormal:
+			if len(e.Data) == 0 {
+				// Don't drop empty proposals. We don't really expect those to come in from
+				// remote nodes (as they're proposed by new leaders), but it does happen.
+				// Whether these are dropped or not should not matter. We opt to not drop
+				// them.
+				return false
+			}
 			cmdID, _ := DecodeRaftCommand(e.Data)
 			if _, ok := r.mu.remoteProposals[cmdID]; !ok {
 				// Untracked remote proposal. Don't drop.
@@ -4047,6 +4054,11 @@ func (r *Replica) maybeTrackForwardedProposalLocked(rg *raft.RawNode, req *RaftM
 	for _, e := range req.Message.Entries {
 		switch e.Type {
 		case raftpb.EntryNormal:
+			if len(e.Data) == 0 {
+				// Ignore empty proposals, which are different than proposals
+				// with no data. These are sent on leadership changes.
+				continue
+			}
 			cmdID, data := DecodeRaftCommand(e.Data)
 			if len(data) == 0 {
 				// An empty command is proposed to unquiesce a range and

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -8395,7 +8395,7 @@ func TestReplicaShouldDropForwardedProposal(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	cmdSeen, cmdNotSeen := makeIDKey(), makeIDKey()
-	data, noData := []byte("data"), []byte("")
+	data := []byte("data")
 
 	testCases := []struct {
 		name                string
@@ -8404,6 +8404,30 @@ func TestReplicaShouldDropForwardedProposal(t *testing.T) {
 		expDrop             bool
 		expRemotePropsAfter int
 	}{
+		{
+			name:   "empty proposal (nil):",
+			leader: true,
+			msg: raftpb.Message{
+				Type: raftpb.MsgProp,
+				Entries: []raftpb.Entry{
+					{Type: raftpb.EntryNormal, Data: nil},
+				},
+			},
+			expDrop:             false,
+			expRemotePropsAfter: 1,
+		},
+		{
+			name:   "empty proposal (zero)",
+			leader: true,
+			msg: raftpb.Message{
+				Type: raftpb.MsgProp,
+				Entries: []raftpb.Entry{
+					{Type: raftpb.EntryNormal, Data: []byte{}},
+				},
+			},
+			expDrop:             false,
+			expRemotePropsAfter: 1,
+		},
 		{
 			name:   "new proposal",
 			leader: true,
@@ -8442,12 +8466,12 @@ func TestReplicaShouldDropForwardedProposal(t *testing.T) {
 			expRemotePropsAfter: 2,
 		},
 		{
-			name:   "empty proposal",
+			name:   "proposal with no data (not an empty proposal)",
 			leader: true,
 			msg: raftpb.Message{
 				Type: raftpb.MsgProp,
 				Entries: []raftpb.Entry{
-					{Type: raftpb.EntryNormal, Data: encodeRaftCommandV1(cmdNotSeen, noData)},
+					{Type: raftpb.EntryNormal, Data: encodeRaftCommandV1(cmdNotSeen, nil)},
 				},
 			},
 			expDrop:             false,

